services/venice-controller/src/main/java/com/linkedin/venice/controller/init/ControllerClientBackedSystemSchemaInitializer.java

```diff
@@ -0,0 +1,246 @@
+package com.linkedin.venice.controller.init;
+
+import static com.linkedin.venice.ConfigKeys.CONTROLLER_SYSTEM_SCHEMA_CLUSTER_NAME;
+
+import com.linkedin.avroutil1.compatibility.AvroCompatibilityHelper;
+import com.linkedin.venice.VeniceConstants;
+import com.linkedin.venice.controller.VeniceHelixAdmin;
+import com.linkedin.venice.controllerapi.ControllerClient;
+import com.linkedin.venice.controllerapi.ControllerResponse;
+import com.linkedin.venice.controllerapi.MultiSchemaResponse;
+import com.linkedin.venice.controllerapi.NewStoreResponse;
+import com.linkedin.venice.controllerapi.SchemaResponse;
+import com.linkedin.venice.controllerapi.StoreResponse;
+import com.linkedin.venice.controllerapi.UpdateStoreQueryParams;
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.exceptions.VeniceNoStoreException;
+import com.linkedin.venice.schema.writecompute.WriteComputeSchemaConverter;
+import com.linkedin.venice.serialization.avro.AvroProtocolDefinition;
+import com.linkedin.venice.utils.Pair;
+import com.linkedin.venice.utils.Utils;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.avro.Schema;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class ControllerClientBackedSystemSchemaInitializer {
+  private static final Logger LOGGER = LogManager.getLogger(ControllerClientBackedSystemSchemaInitializer.class);
+  private static final String DEFAULT_KEY_SCHEMA_STR = "\"int\"";
+  private static final int DEFAULT_RETRY_TIMES = 10;
+
+  private final AvroProtocolDefinition protocolDefinition;
+  private final String clusterName;
+  private final VeniceHelixAdmin admin;
+  private final Schema keySchema;
+  private final UpdateStoreQueryParams storeMetadataUpdate;
+  private final boolean autoRegisterDerivedComputeSchema;
+  private final boolean enforceSSLOnly;
+  private ControllerClient controllerClient;
+
+  public ControllerClientBackedSystemSchemaInitializer(
+      AvroProtocolDefinition protocolDefinition,
+      String systemStoreCluster,
+      VeniceHelixAdmin admin,
+      Schema keySchema,
+      UpdateStoreQueryParams storeMetadataUpdate,
+      boolean autoRegisterDerivedComputeSchema,
+      boolean enforceSSLOnly) {
+    this.protocolDefinition = protocolDefinition;
+    this.clusterName = systemStoreCluster;
+    this.admin = admin;
+    this.keySchema = keySchema;
+    this.storeMetadataUpdate = storeMetadataUpdate;
+    this.autoRegisterDerivedComputeSchema = autoRegisterDerivedComputeSchema;
+    this.enforceSSLOnly = enforceSSLOnly;
+  }
+
+  public void execute() {
+    String storeName = protocolDefinition.getSystemStoreName();
+    Map<Integer, Schema> schemasInLocalResources = Utils.getAllSchemasFromResources(protocolDefinition);
+
+    String leaderControllerUrl = admin.getLeaderController(clusterName).getUrl(enforceSSLOnly);
```

Comment:
Do we need to worry too much about the leader transitions during the time when `execute()` is in execution, the controllerClient that has been cached could be stale?

Reply:
Don't worry. Every request will call `getLeaderControllerUrl()` to get the up-to-date leader controller url: https://github.com/linkedin/venice/blob/bfad45829cf9224b3053c1682d5b0780aa98f7b6/internal/venice-common/src/main/java/com/linkedin/venice/controllerapi/ControllerClient.java#L1227